### PR TITLE
Add request coalescing and relevant configuration

### PR DESCRIPTION
**Pull request by fuatbasik**
## Description of change
With this change, AAL can coalesce requests in a close proximity to each other. This enables to take spatial locality
into account and reduce request count. In my testing I observed that this is particularly important for vectored reads
reading vectors near each other. 

#### Relevant issues
N/A

#### Does this contribution introduce any breaking changes to the existing APIs or behaviors?
No. It changes the behaviour of vectored reads and merges smaller requests to bigger requests.
These bigger requests might be split to smaller requests later to make sure all requests are around `target.request.size`

#### How was the contribution tested?
Added new unit tests. Extended existing IntegrationTests to run for both coalescing enabled and disabled. 

#### Does this contribution need a changelog entry?
- Yes, i shall be updating the README after we finalise the behaviour part of this PR. 

---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and I agree to the terms of the [Developer Certificate of Origin (DCO)](https://developercertificate.org/).

### CODE REVIEW

**Review by ahmarsuhail — CHANGES_REQUESTED** [review]
Looks good, just think we need to add some additional logic in that execute function

File: input-stream/src/integrationTest/java/software/amazon/s3/analyticsaccelerator/access/ReadVectoredTest.java [hunk 1]
```diff
@@ -391,12 +404,12 @@ protected void testReadVectoredInSingleBlock(
   protected void testReadVectoredForSequentialRanges(
```
[the remark this answers is not in the document]
Comment (fuatbasik) [reply input-stream/src/integrationTest/java/software/amazon/s3/analyticsaccelerator/access/ReadVectoredTest.java:404]: Thanks @ahmarsuhail. Do test under IOPlanTest cover your concern. I was trying to put corner cases for actual coalesing business logic there instead of here.

File: input-stream/src/main/java/software/amazon/s3/analyticsaccelerator/io/physical/plan/IOPlan.java [hunk 2]
```diff
@@ -58,4 +58,33 @@ public String toString() {
...
+      // Always compare with the last checked interval to avoid n^2 comparisons
+      if (currentRange.getEnd() + tolerance >= nextRange.getStart()) {
+        currentRange =
+            new Range(currentRange.getStart(), Math.max(currentRange.getEnd(), nextRange.getEnd()));
+      } else {
+        coalescedRanges.add(currentRange);
+        currentRange = nextRange;
+      }
+    }
+    coalescedRanges.add(currentRange);
```
[the remark this answers is not in the document]
Comment (fuatbasik) [reply input-stream/src/main/java/software/amazon/s3/analyticsaccelerator/io/physical/plan/IOPlan.java:86]: Thanks. I updated this part in rev2

**Review by ozkoca — DISMISSED** [review]
LGTM

File: input-stream/src/main/java/software/amazon/s3/analyticsaccelerator/io/physical/impl/PhysicalIOImpl.java [hunk 3]
```diff
@@ -213,10 +218,28 @@ public int readTail(byte[] buf, int off, int len) throws IOException {
...
+  /**
+   * Async method capable of executing a logical IO plan with request coalesce flag.
+   *
+   * @param ioPlan the plan to execute asynchronously
+   * @param readMode the read mode for which this IoPlan is being executed
+   * @param coalesceRequests flag to enable request coalescing
+   * @return an IOPlanExecution object tracking the execution of the submitted plan
+   */
+  public IOPlanExecution execute(IOPlan ioPlan, ReadMode readMode, boolean coalesceRequests) {
+    if (coalesceRequests) {
```
Comment (ahmarsuhail) [inline input-stream/src/main/java/software/amazon/s3/analyticsaccelerator/io/physical/impl/PhysicalIOImpl.java:238]: sorry for asking for another change, but I realised that a better way to do this is using the enum. In the enum, for all the readModes we have, you can add a new method which will return true or false depending on if we want to coalesce or not. do you think it's worth making that change?

File: input-stream/src/main/java/software/amazon/s3/analyticsaccelerator/io/physical/plan/IOPlan.java [hunk 4]
```diff
@@ -58,4 +58,33 @@ public String toString() {
...
+   * Coalesces the prefetch ranges in this plan. The coalescing is done in-place.
+   *
+   * @param tolerance the tolerance for coalescing
+   */
+  public void coalesce(long tolerance) {
+    if (this.prefetchRanges.size() < 2) {
+      return;
+    }
+    // Ensure ranges are ordered by their start position.
+    Collections.sort(this.prefetchRanges);
```
[the remark this answers is not in the document]
Comment (ahmarsuhail) [reply input-stream/src/main/java/software/amazon/s3/analyticsaccelerator/io/physical/plan/IOPlan.java:72]: similarly, in the readMode you can have is this list already sorted or not (eg: readVectored will always be sorted). 

i know that these lists are tiny and it doesn't matter if you sort it twice, but just wondering if that might be a better approach

**Review by ahmarsuhail — APPROVED** [review]
LGTM

File: input-stream/src/main/java/software/amazon/s3/analyticsaccelerator/io/physical/impl/PhysicalIOImpl.java (hunk 3 above)
Comment (fuatbasik) [reply input-stream/src/main/java/software/amazon/s3/analyticsaccelerator/io/physical/impl/PhysicalIOImpl.java:238]: makes sense. I am adding it now.

**Review by ahmarsuhail — CHANGES_REQUESTED** [review]
typo in the coalesce method, using the wrong variable
